clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.common.message.ConsumerGroupHeartbeatResponseData;
+import org.apache.kafka.common.protocol.Errors;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Membership manager that maintains group membership for a single member following the new
+ * consumer group protocol.
+ * <p/>
+ * This keeps membership state and assignment updated in-memory, based on the heartbeat responses
+ * the member receives. It is also responsible for computing assignment for the group based on
+ * the metadata, if the member has been selected by the broker to do so.
+ */
+public class MembershipManagerImpl implements MembershipManager {
+
+    private final String groupId;
+    private Optional<String> groupInstanceId;
+    private String memberId;
+    private int memberEpoch;
+    private MemberState state;
+    private AssignorSelection assignorSelection;
+
+    /**
+     * Assignment that the member received from the server and successfully processed
+     */
+    private ConsumerGroupHeartbeatResponseData.Assignment currentAssignment;
+
+    /**
+     * List of assignments that the member received from the server but hasn't processed yet
+     */
+    private final List<ConsumerGroupHeartbeatResponseData.Assignment> targetAssignments;
```

Comment:
Yes, your understanding is right. We need to keep a single target assignment, but since we'll continue to heartbeat while we reconcile, we also need to keep newer assignments that we might receive while reconciling the 1st one (need to keep only the last one though, as you pointed out). 

Maybe it'll make more clear if I keep something like a single `targetAssignment` (potentially in the process of being reconciled), and a single  optional `nextTargetAssignment`. In the end, I don't expect that this list will ever have more than 2 items.

Reply:
I would prefer having three attributes personally because it is clearer.

Reply:
I think that we need to think a little more about how we will handle this. My concern is that the server side may start timers based on the `nextTargetAssignment` but the client won't look into it before it finishes the reconciliation to the `targetAssignment`.

For instance, imagine the following scenario.
* Current assignment is A and B;
* Server asks to revoke A and starts the rebalance timer (5 minutes by default);
* Client starts the revocation of A. This can take a while;
* Server asks to revoke B and restarts the rebalance timer;
* Client still waits on the revocation of A to complete. When completed, it starts the revocation of B.

In this scenario, if the revocation of A and B takes more than 5 minutes altogether, the member will be kicked out from the consumer group. The server does not know that the consumer is blocked on the revocation of A.

Do we really need to wait until A is fully revoked before we can start the process to revoke B?

Reply:
Very interesting, we've had some conversations about this (without clear ideas yet) so I need to think about this more. In principle I don't see a reason why we couldn't process the 2 revocations above in parallel, given they relate to different partitions. It feels like the right thing to do, to take advantage of the incremental assignment when possible. That being said, it would definitely be a change in how we execute the callbacks. User's callbacks could be executed concurrently for distinct sets of partitions, which is something that does not happen today (they are invoked sequentially in the poll loop)

I will update this initial code to have the clearer var definitions for now: `currentAssignment`, `targetAssignment` and `nextTargetAssignment`. I will continue to think about how to handle overlapping assignments, which seems to be where we could consider the trade-offs of starting processing a `nextTargetAssignment` even if there is a `targetAssignment` being processed.

Reply:
> User's callbacks could be executed concurrently for distinct sets of partitions, which is something that does not happen today (they are invoked sequentially in the poll loop)

This does not seem doable to me because it would break the contract of the `poll` method, isn't it?

The alternative that I am considering would be to not let the server compute a new assignment until the previous one is ack'ed. I need to think this though a little more. The advantage is that it simplifies the client side (we don't need nextTargetAssignment). The disadvantage is that it is a bit less incremental.

Reply:
Agree that the actual process on the client should be sync to maintain the current contract. It would be only when receiving a new assignment and having already a nextTarget, that we could get smarted on the client side (Ex. nextTargetAssignment being a revocation of partition A, and receiving a revocation of partition B, could we maybe bundle them together on the client, before they are processed)

Reply:
the server side change you have in mind seems the cleanest to reason about of course, but agree the it's a bit less incremental